db/kv_checksum.cc

```diff
@@ -0,0 +1,155 @@
+//  Copyright (c) 2020-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/kv_checksum.h"
+
+#include <type_traits>
+
+namespace ROCKSDB_NAMESPACE {
+
+namespace {
+
+// bitwise operations on an enum class adapted from
+// https://softwareengineering.stackexchange.com/a/204566
+KvProtectionInfo::CoverageFlags operator|(KvProtectionInfo::CoverageFlags lhs,
+                                          KvProtectionInfo::CoverageFlags rhs) {
+  using T = std::underlying_type<KvProtectionInfo::CoverageFlags>::type;
+  return static_cast<KvProtectionInfo::CoverageFlags>(static_cast<T>(lhs) |
+                                                      static_cast<T>(rhs));
+}
+
+KvProtectionInfo::CoverageFlags& operator|=(
+    KvProtectionInfo::CoverageFlags& lhs, KvProtectionInfo::CoverageFlags rhs) {
+  lhs = lhs | rhs;
+  return lhs;
+}
+
+KvProtectionInfo::CoverageFlags operator&(KvProtectionInfo::CoverageFlags lhs,
+                                          KvProtectionInfo::CoverageFlags rhs) {
+  using T = std::underlying_type<KvProtectionInfo::CoverageFlags>::type;
+  return static_cast<KvProtectionInfo::CoverageFlags>(static_cast<T>(lhs) &
+                                                      static_cast<T>(rhs));
+}
+
+}  // anonymous namespace
+
+KvProtectionInfo::KvProtectionInfo() { Clear(); }
+
+Status KvProtectionInfo::VerifyAgainst(const KvProtectionInfo& expected) const {
+  if (expected.HasKeyChecksum() &&
+      GetKeyChecksum() != expected.GetKeyChecksum()) {
+    return Status::Corruption("key checksum mismatch");
+  }
+  if (expected.HasValueChecksum() &&
+      GetValueChecksum() != expected.GetValueChecksum()) {
+    return Status::Corruption("value checksum mismatch");
+  }
+  if (expected.HasSequenceNumber() &&
+      GetSequenceNumber() != expected.GetSequenceNumber()) {
+    return Status::Corruption("sequence number mismatch");
+  }
+  if (expected.HasColumnFamilyId() &&
+      GetColumnFamilyId() != expected.GetColumnFamilyId()) {
+    return Status::Corruption("column family ID mismatch");
+  }
+  if (expected.HasValueType() && GetValueType() != expected.GetValueType()) {
+    return Status::Corruption("value type mismatch");
+  }
+  return Status::OK();
+}
+
+void KvProtectionInfo::Clear() {
+  coverage_flags_ = static_cast<CoverageFlags>(0);
```

Comment:
There is a potential issue here if you are running with asserts off.  In that case, the values are not initialized, meaning they are random.  Or if you called Clear on an object that had previously been used, the values will match the old settings.  So if you call GetXXX (e.g, GetChecksum), it will return a random or old value if asserts are off.

Seems like you need to set the values to some default (or illegal value) here.

Reply:
This code is removed.